pkg/controller/node_failure_controller/node_failure_controller.go

```diff
@@ -0,0 +1,406 @@
+	if node.Status.Phase == v1.NodeTerminated {
+		return true
+	}
+	objName, err := controller.KeyFunc(node)
+	if err != nil {
+		glog.Errorf("failed to get key from object: %v", err)
+		watcher.enqueueWork(watcher.nodeQueue, node)
+		return false
+	}
+
+	for _, condition := range node.Status.Conditions {
+		if condition.Type == v1.NodeReady && condition.Status != v1.ConditionTrue {
+			now := time.Now()
+			firstMarkTime, ok := watcher.nodeFirstBrokenMap[objName]
+			if ok {
+				timeInterval := now.Sub(firstMarkTime)
+				if timeInterval.Seconds() > DefaultNodeNotReadyTimeDuration.Seconds() {
+					return true
+				} else {
+					glog.V(6).Infof("node:%s is not ready, but less than 2 minutes, re-enqueue", node.Name)
+					// NotReady status lasts less than 2 minutes
+					// re-enqueue
+					watcher.enqueueWork(watcher.nodeQueue, node)
+					return false
+				}
+			} else {
+				// first time to mark the node NotReady
+				watcher.nodeFirstBrokenMap[objName] = now
+				watcher.enqueueWork(watcher.nodeQueue, node)
+				return false
+			}
+		}
+	}
+
+	// The node status is ok, but if it was marked before, remove the mark
+	_, ok := watcher.nodeFirstBrokenMap[objName]
+	if ok {
+		delete(watcher.nodeFirstBrokenMap, objName)
+	}
+	return false
+}
+
+func (watcher *NodeWatcher) deleteNode(key string, node *v1.Node) {
+	glog.Infof("node:%s is deleted, so mark the local PVs on it", node.Name)
+	watcher.nodeMap.DeleteNode(key)
+
+	// mark all local PVs on this node
+	// try several times again
+	for i := 0; i < local_pv_monitor.UpdatePVRetryCount; i++ {
+		err := watcher.markLocalPVsOnNode(node, NodeFailure, "true")
+		if err != nil {
+			glog.V(4).Infof("marking local PVs failed: %v", err)
+			time.Sleep(local_pv_monitor.UpdatePVInterval)
+			continue
+		}
+		return
+	}
+
+	// when we reach here, means that marking local failed, re-enqueue
+	watcher.enqueueWork(watcher.nodeQueue, node)
+}
+
+func (watcher *NodeWatcher) markLocalPVsOnNode(node *v1.Node, ann, val string) error {
+	pvs, err := watcher.volumeLister.List(labels.NewSelector())
+	if err != nil {
+		glog.Warningf("cannot list pvs: %s", err)
+		return err
+	}
+
+	var markErrHappened bool
+	for _, pv := range pvs {
+		// if the pv is not local pv, skip
+		if pv.Spec.Local == nil {
+			continue
+		}
+		// check node and pv affinity
+		fit, err := local_pv_monitor.CheckNodeAffinity(pv, node.Labels)
+		if err != nil {
+			glog.Errorf("check node affinity error: %v", err)
+			return err
+		}
+		if !fit {
+			glog.Errorf("pv: %s does not belong to this node: %s", pv.Name, node.Name)
+			continue
+		}
+
+		err = watcher.markPV(pv, ann, val)
+		if err != nil {
+			markErrHappened = true
+			glog.Errorf("mark local PV: %s failed, error: %v", pv.Name, err)
+		}
+	}
+
+	if markErrHappened {
+		return fmt.Errorf("error happened when marking local PVs")
+	}
+	return nil
+}
+
+func (watcher *NodeWatcher) markPV(pv *v1.PersistentVolume, ann, value string) error {
```

Comment:
is this duplicate with markPV from local-pv-monitor?

Reply:
yes, most of them are the same, but we need to send record events and use client here, so i re-write it. We can revisit this later

Reply:
fix this in https://github.com/caicloud/kube-storage-monitor/pull/11